ELEC-98: Add Vagrantfile for uwmidsun/box

karlding opened the pull request:
So I think I've finally managed to figure out the reason why USB passthrough on macOS was failing. There's actually a [regression in VirtualBox](https://www.virtualbox.org/ticket/15956) that causes the STLink programmer to not show up. This afternoon, I downgraded to **VirtualBox 5.0.8**, and then installed the appropriate Extension Pack on my MacBook, and it seems to work. I've added that to the installation instructions in the ``README.md``.

There's also currently an issue with ``vb.customize``, where it will add a new ``usbfilter`` whenever ``vagrant up`` or ``vagrant reload`` is run. I can fix this with some Ruby (since the ``Vagrantfile`` is just Ruby) that checks whether the ``usbfilter`` already exists, but I'm not sure whether it'll be portable enough to handle different permutations of where ``VBoxManage`` is located on Windows (like, is it always available on the ``PATH``?).

I've tested this on macOS and Ubuntu 16.04 LTS.

I won't merge this until we upload the box to Atlas (or somewhere else), but this is just for a preliminary review. In the meantime, you'd just have to do ``vagrant box add uwmidsun/box /path/to/vagrant-box.box``

karlding committed:
ELEC-98: Initial commit

Add working Vagrantfile and install instructions.

karlding committed:
ELEC-98: Add instructions for macOS

Add installation instructions for macOS regarding regression.

karlding commented:
Also, we should probably figure out what installation permutations to support on Windows, so we can provide more specific cases that we support (ie. do we force everyone to use Git Bash, or PuTTY?) on Confluence.

karlding committed:
ELEC-98: Fix for macOS VirtualBox 5.1 regression

Fix for the networking issue present in VirtualBox 5.1

karlding commented:
Also, unfortunately, downgrading VirtualBox for macOS results in another regression. So we need to add

```ruby
config.vm.provider 'virtualbox' do |vb|
  vb.customize ["modifyvm", :id, "--cableconnected1", "on"]
end
```

to the ``Vagrantfile``.

ckitagawa commented:
I'll test this out on Windows before Thursday's meeting. Thanks Karl! I would argue for Git Bash, more useful in general, good to learn, and less of a hassle than PuTTY IMO.

karlding commented:
Regarding the issue with duplicate filters, should I go ahead and merge my current code for testing? It works, although my main concern with it is regarding how portable it will be, considering Windows is notorious for issues with executables on the ``PATH`` (since my code relies on the ``VBoxManage`` command).

karlding committed:
ELEC-98: Update README instructions

Update README to include vagrant commands, and update USBFilter
setup instructions

karlding committed:
ELEC-98: Update README

Update the README with Windows instructions to enable Virtualization
in BIOS

karlding commented:
Hmm, now that we're adding a requirement for users to have ``VBoxManage`` on their ``PATH``, I'll just merge in my Ruby code for better USBFilter functionality. In the future, that can be extended to perhaps detect where ``VBoxManage.exe`` is located.

karlding commented:
I talked with Aaron and set up an Atlas account for our organization, so I'll just build a new image using ``packer`` and then upload the box there, so we can actually test using ``vagrant up`` instead of running ``vagrant box add uwmidsun/box`` and then ``vagrant up``.

karlding commented:
I've released the box on [Atlas](https://atlas.hashicorp.com/uwmidsun/boxes/box/versions/1.0.0), so we should be able to just clone and then follow the instructions in the **README.md** file now.

karlding commented:
I've been thinking about the best way to approach the duplicate usbfilter issue, and I think the way to go is basically similar to progressive enhancement: 

* If they have ``VBoxManage`` on the ``PATH``, then use it 
* Otherwise, we'll just insert a duplicate usbfilter (if they care, they can add it to the ``PATH``)

This approach should be doable without needing to resort to OS detection, or forcing people to add things to their ``PATH``, and should be more reliable than guessing where Windows users installed VirtualBox.

karlding committed:
ELEC-98: Add a better usbfilter wrapper

Add a working usbfilter wrapper and Which implementation

karlding commented:
@ckitagawa I implemented the ``usbfilter`` fix. I'm open to put the dependency in the ``BetterUSB`` module, instead of leaving it in the ``Vagrantfile``.

ckitagawa commented on README.md:
```diff
@@ -1,2 +1,116 @@
+
+The ARM toolchain that will be set up includes the following tools:
+
+* GCC ARM Embedded
```
The Commonmark md style guides I've seen call for three spaces like:
```*···Words```

I honestly don't care that much just a suggestion.

ckitagawa reviewed (approved):
LGTM, Approval.

karlding committed:
ELEC-98: Update README with Windows instructions

Update README with Windows installation instructions
